common/src/main/java/io/druid/math/expr/Parser.java

```diff
@@ -70,7 +93,51 @@ public static Expr parse(String in)
     ParseTreeWalker walker = new ParseTreeWalker();
     ExprListenerImpl listener = new ExprListenerImpl(parseTree);
     walker.walk(listener, parseTree);
-    return listener.getAST();
+    return withFlatten ? flatten(listener.getAST()) : listener.getAST();
+  }
+
+  public static Expr flatten(Expr expr)
+  {
+    if (expr instanceof BinaryOpExprBase) {
+      BinaryOpExprBase binary = (BinaryOpExprBase) expr;
+      Expr left = flatten(binary.left);
+      Expr right = flatten(binary.right);
+      if (Evals.isAllConstants(left, right)) {
+        expr = expr.eval(null).toExpr();
+      } else if (left != binary.left || right != binary.right) {
+        return Evals.binaryOp(binary, left, right);
+      }
+    } else if (expr instanceof UnaryExpr) {
+      UnaryExpr unary = (UnaryExpr) expr;
+      Expr eval = flatten(unary.expr);
+      if (Evals.isConstant(eval)) {
+        expr = expr.eval(null).toExpr();
+      } else if (eval != unary.expr) {
+        if (expr instanceof UnaryMinusExpr) {
+          expr = new UnaryMinusExpr(eval);
+        } else if (expr instanceof UnaryNotExpr) {
+          expr = new UnaryNotExpr(eval);
+        } else {
+          expr = unary; // unknown type..
+        }
+      }
+    } else if (expr instanceof FunctionExpr) {
+      FunctionExpr functionExpr = (FunctionExpr) expr;
+      List<Expr> args = functionExpr.args;
+      boolean flattened = false;
+      List<Expr> fattening = Lists.newArrayListWithCapacity(args.size());
+      for (int i = 0; i < args.size(); i++) {
+        Expr flatten = flatten(args.get(i));
+        flattened |= flatten != args.get(i);
+        fattening.add(flatten);
+      }
+      if (Evals.isAllConstants(args)) {
+        expr = expr.eval(null).toExpr();
```

Comment:
I'm confused, what's going on here? It looks like this is saying that `f(const) = null` which doesn't seem right.